src/dmd/dinterpret.d

```diff
@@ -6095,6 +6095,30 @@ public:
             se = (cast(ClassReferenceExp)ex).value;
             i = (cast(ClassReferenceExp)ex).findFieldIndexByName(v);
         }
+        else if (ex.op == TOK.typeid_)
+        {
+            auto tie = cast(TypeidExp) ex;
+            if (v.ident == Identifier.idPool("name"))
+            {
+                if (auto t = isType(tie.obj))
+                {
+                    auto sym = t.toDsymbol(null);
+                    auto ident = (sym ? sym.ident : null);
+                    if (ident)
+                    {
+                        result = new StringExp(e.loc, ident.toString());
+                        result.expressionSemantic(null);
+                        return ;
+                    }
+                }
+            }
+            // We get here if we could not resolve the name or we were not asked for the name
+            {
+                e.error("only typeid(class/struct).name is implemented at compile time");
```

Comment:
I would rephrase the error message to something along the lines of: "At compile time, `typeid` works only if the provided type is `class` or `struct` and the `name` property is accessed: `typeid(class/struct).name`"

Reply:
I don't really care.

Reply:
`typeid(class/struct).name` is a somewhat ambiguous message. Obviously typeid(class) and typeid(struct) doesn't work. But...

1. Is this going to support typeid(C) where C is a class or struct type?
2. The message isn't clear enough that it supports class instance and struct instance typeid
3. I'm not sure why we need to restrict this to only classes and structs. Is there a technical reason why typeid(int) or typeid(i) (where i is an int) doesn't work?

Ideally `typeid(anything).name` works for everything the same as it does at runtime. Then code doesn't have to be written specifically for ctfe mode.

A further question I just thought of out of ignorance. Is the result of this going to be identical to typeid(x).name during runtime? If not, then it needs to be.

Reply:
the reason typeid(basic_type).name is not supported is that it would take more code to do it.

Reply:
Having an error message that is informative to the user is crucial to the acceptance of the PR.